### PR TITLE
gpstate: add "Rewinding" status for primary during incremental recovery

### PR DESCRIPTION
**Pull request by jchampio**
This commit checks whether there are any active pg_rewind connections when the mirror is down and displays an appropriate status message if so.  To facilitate this, the pg_rewind call in gprecoverseg has been modified to specify its application_name when connecting to the database.

## Here are some reminders before you submit the pull request
- [x] Add tests for the change
- [ ] ~Document changes~ (not yet)
- [x] Pass `make installcheck`

## Screenshots
image

### CODE REVIEW

File: gpMgmt/bin/gppylib/commands/gp.py [hunk 1]
```diff
@@ -407,7 +407,9 @@ def __init__(self, name, target_host, target_datadir,
                  verbose=False, ctxt=REMOTE):
 
         # Construct the source server libpq connection string
-        source_server = "host=%s port=%s dbname=template1" % (source_host, source_port)
+        # We set application_name here so gpstate can identify whether an
+        # incremental recovery is occurring.
+        source_server = "host=%s port=%s dbname=template1 application_name=pg_rewind" % (source_host, source_port)
```
Comment (ashwinstar) [inline gpMgmt/bin/gppylib/commands/gp.py]: Management scripts today do very bad job at keeping the things in one place, like the name application_name `pg_rewind`. I would really like it to be defined in one place and used wherever is used for easy management in future.

Also, pg_rewind seems very generic name (yes we don't document pg_rewind as tool today). But seems better would be to use some GPDB specific internal name for this to make sure doesn't get confused by any other instance of pg_rewind like maybe atleast `intetnal_pg_rewind` or `gprecoverseg_pg_rewind` or `incremental_recovery`.

File: gpMgmt/bin/gppylib/programs/clsSystemState.py [hunk 2]
```diff
@@ -891,19 +892,32 @@ def _add_replication_info(data, primary, mirror):
...
         # Successfully queried pg_stat_replication. If there are any backup
-        # connections, mention them in the primary status.
+        # or pg_rewind connections, mention them in the primary status.
+        state = None
         backup_connections = [r for r in rows if r[1] == 'backup']
         if backup_connections:
             row = backup_connections[0]
+            state = replication_state_to_string(row[1])
+        elif rewinding:
+            state = "Rewinding history to match primary timeline"
```
Comment (ashwinstar) [inline gpMgmt/bin/gppylib/programs/clsSystemState.py]: maybe "Incrementally recovering primary"?
I feel rewind is too little heavy concept to explain as state to users and maybe not required. As anyways it just doens't do rewinding but also fast forward by copying current xlog files. Unless we really wish to be specific than abstract, I am okay keeping what we currently have.

Comment (jchampio) [reply gpMgmt/bin/gppylib/programs/clsSystemState.py]: > As anyways it just doens't do rewinding but also fast forward by copying current xlog files.

Is this true? My understanding was that pg_rewind only took the WAL back to the point of divergence; it's up to the mirror to play catchup once it starts.

File: gpMgmt/bin/gppylib/commands/gp.py (hunk 1 above)
Comment (jchampio) [reply gpMgmt/bin/gppylib/commands/gp.py]: Pushed an improvement, thanks!

File: gpMgmt/bin/gppylib/programs/clsSystemState.py (hunk 2 above)
Comment (jimmyyih) [reply gpMgmt/bin/gppylib/programs/clsSystemState.py]: Double checked and yes, it is true.  The pg_rewind tool will rewind the target to the divergence point, sync up the WAL and files, and then update the pg_control file.  So with that knowledge, yeah the statement we're outputting is a bit deceiving.

File: gpMgmt/bin/gppylib/programs/clsSystemState.py [hunk 3]
```diff
@@ -891,19 +892,32 @@ def _add_replication_info(data, primary, mirror):
...
             logger.warning('could not query segment {} ({}:{})'.format(
                     primary.dbid, primary.hostname, primary.port
             ), exc_info=ie)
             return
 
         # Successfully queried pg_stat_replication. If there are any backup
-        # connections, mention them in the primary status.
+        # or pg_rewind connections, mention them in the primary status.
+        state = None
         backup_connections = [r for r in rows if r[1] == 'backup']
```
Comment (slari-pivotal) [inline gpMgmt/bin/gppylib/programs/clsSystemState.py]: If `rewinding` is true, would we still want to compute `backup_connections`?  If not, then this line can go in `if !rewinding:` case.

Comment (jchampio) [reply gpMgmt/bin/gppylib/programs/clsSystemState.py]: For now, backup takes precedence over rewind. This is an unanswered UX issue at the moment -- what do we display if both are true? -- but for now, this is as designed.

**Comment by jchampio** [comment]
Rebased onto master; there were a couple of conflicting changes from #6897 that needed to be resolved.

**Pull request merged**